woocommerce/woocommerce-gutenberg-products-block
# Make stylesheets order consistent between build and dev

Fixes #2713.

### How to test the changes in this Pull Request:

First, in the main branch, let's write some styles to make the issue visible:
* In [`assets/js/base/components/payment-methods/style.scss`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/assets/js/base/components/payment-methods/style.scss), add these styles:
```CSS
body {
	border: 4px solid red;
}
```
* In [`assets/js/blocks/cart-checkout/cart/style.scss`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/assets/js/blocks/cart-checkout/cart/style.scss), add these styles:
```CSS
body {
	border: 4px solid blue;
}
```
1. Run `npm run start` and see the body has a blue border.
2. Run `npm run build` and notice the body border is now red. Styles are different between start and build! :scream: 

Now, checkout this branch (`fix/2713-stylesheets-build-order`):
1. Run `npm run start` and see the body border is still blue.
2. Run `npm run build` and notice the body border is now blue. Styles are the same between start and build! :tada: 

In addition to that, it would be good to test that there are no regressions when running `npm run package plugin` and `npm run storybook`.

### Explanation of the changes

#### How the CSS is currently generated:

Even though we have several [webpack configs](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/bin/webpack-configs.js#L537-L540) (`core`, `main`, `front` and `payments`), all CSS files are generated from the `main` config.

In most cases, that worked fine: `main` has all blocks' `index.js` as the entry point, those files were in charge of importing `style.scss`, `editor.scss` and the components in use, so all required styles were caught by webpack and later bundled in `style.css`, `editor.css`.

However, there were some cases where this config failed: for example, legacy builds didn't have the All Products block, so they were missing the product list styles. To fix that, the entry point [was added manually](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/bin/webpack-entries.js#L26-L31) (see #2000).

Something similar happened with the BlockErrorBoundary styles: that component was only used in frontend scripts, so was not caught by `main` entries. Because of that, it had to be [added manually too](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/bin/webpack-entries.js#L58-L59). Some time later, the BlockErrorBoundary has been imported from scripts that are loaded in the `main` config, so it's no longer necessary to explicitly declare that SCSS file in the entries list. However, we didn't notice it so it's still there.

The current approach, has another issue, the one that is causing #2713. Since we started using tree shaking, styles order stopped being something reliable and it depended on the order components and blocks were imported. That's why `npm run start` and `npm run build` generate the CSS with a different order. We were basically facing this issue: https://github.com/webpack/webpack/issues/7094.

#### How the CSS is generated with this pull:

In order to fix the issues above and (hopefully) simplify the way our CSS is generated, I split the CSS generation config from `main` into its own config (`styling`). It declares the CSS files as entries, this guarantees styles are always included in the same order.

#### Pros of this approach:
* CSS order is consistent between dev and prod.
* All CSS files under `assets` are included, so there is no need to manually add `scss` files that are not imported from `main` config.
* Some editor styles that leaked into `style.css` with the previous approach, are no longer there, producing a 6% decrease in file size for `style.css`.

#### Cons of this approach:

* With this approach, there is no way to know which component styles should be bundled or not. If a component is only used in an experimental block, its styles will still be bundled in the final `style.css`. Same for legacy builds: they have the same component styles as regular builds. You will notice that means `legacy` stylesheets size increased a lot.

I know this is not a perfect solution. It solves some issues while introducing some cons. I would like to get some feedback whether the team feels this is worth merging or somebody has another idea. :slightly_smiling_face:
